### PR TITLE
Avx512/rol ror

### PR DESCRIPTION
**Pull request by jtlap**

### CODE REVIEW

File: include/eve/function/rotl.hpp [hunk 1]
```diff
@@ -22,11 +23,14 @@ namespace eve
...
     {
-      EVE_ASSERT( assert_good_shift<T>(s),
+      constexpr int l [[maybe_unused]] = sizeof(value_type_t<T>) * 8;
+      EVE_ASSERT( assert_good_shift<T>(eve::abs(s)),
                   "[eve::rotl] Rotating by " << s
-                                            << " is out of the range [0, "
-                                            << sizeof(value_type_t<T>) * 8
-                                            << "[."
+                                             << " is out of the range ]"
+                                             << -l
```
Comment (jfalcou) [inline include/eve/function/rotl.hpp:30]: why the change here ?

File: include/eve/function/rotr.hpp [hunk 2]
```diff
@@ -19,14 +20,17 @@ namespace eve
...
     {
-      EVE_ASSERT( assert_good_shift<T>(s),
+      constexpr int l [[maybe_unused]] = sizeof(value_type_t<T>) * 8;
+      EVE_ASSERT( assert_good_shift<T>(eve::abs(s)),
                   "[eve::rotr] Rotating by " << s
-                                            << " is out of the range [0, "
-                                            << sizeof(value_type_t<T>) * 8
-                                            << "[."
+                                             << " is out of the range ]"
+                                             << -l
```
Comment (jfalcou) [inline include/eve/function/rotr.hpp:30]: Same here

File: include/eve/module/real/core/function/regular/generic/rotl.hpp [hunk 3]
```diff
@@ -19,16 +19,34 @@
...
-  template<unsigned_value T, integral_value U>
-  [[nodiscard]] EVE_FORCEINLINE auto rotl_(EVE_SUPPORTS(cpu_), T a0, U n) noexcept
+  template<unsigned_scalar_value T, integral_scalar_value U>
+  [[nodiscard]] EVE_FORCEINLINE auto rotl_(EVE_SUPPORTS(cpu_), T a0, U nn) noexcept
   {
-    if constexpr( has_native_abi_v<T> && scalar_value<U> )
+    if( nn == 0 )  return a0;
+    constexpr U width = sizeof(T) * 8 - 1;
+    int n = nn & width;
+    //   std::cout << "icotte" << std::endl;
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/generic/rotl.hpp:28]: remove don"t comment

File: include/eve/module/real/core/function/regular/generic/rotl.hpp [hunk 4]
```diff
@@ -40,13 +58,30 @@ namespace eve::detail
...
-    // now none is scalar
-    else if constexpr( has_emulated_abi_v<T> || has_emulated_abi_v<U> )
+   else
+    {
+      return apply_over(rotl, a0, n);
+    }
+  }
+
+  template<unsigned_simd_value T, integral_simd_value U>
+  [[nodiscard]] EVE_FORCEINLINE auto rotl_(EVE_SUPPORTS(cpu_), T a0, U n) noexcept
```
Comment (jfalcou) [inline include/eve/module/real/core/function/regular/generic/rotl.hpp:82]: why the random [[nodiscard]] ?

File: test/unit/module/real/core/rotl.cpp [hunk 5]
```diff
@@ -57,20 +58,27 @@ auto maxi = []<typename T>(eve::as<T> const & )
 };
 
 //==================================================================================================
-// rotl  tests
+//== rotl  tests
 //==================================================================================================
 EVE_TEST( "Check behavior of rotl on wide"
-        , eve::test::simd::unsigned_integers
+        , eve::test::simd::uints_16//unsigned_integers
```
Comment (jfalcou) [inline test/unit/module/real/core/rotl.cpp:64]: why the change ?

File: test/unit/module/real/core/rotl.cpp [hunk 6]
```diff
@@ -57,20 +58,27 @@ auto maxi = []<typename T>(eve::as<T> const & )
...
+//   TTS_EQUAL( rotl(a0, 0u), a0);
+//   TTS_EQUAL( rotl(a0, 1u), map([](auto e){return std::rotl(e, 1u);}, a0));
   using v_t = eve::element_type_t<T>;
+  a0 = eve::one(eve::as(a0));
+  a1 = eve::detail::linear_ramp(eve::as(a0));
+  a1 = a1 % (sizeof(v_t)*8);
   auto ua1 = eve::int_(a1);
-  TTS_EQUAL( rotl(a0, a1), map([](auto e, auto f) -> v_t {return std::rotl(e, f);}, a0, ua1));
+  std::cout << "a0 " << a0 << std::endl;
+  std::cout << "a1 " << a1 << std::endl;
```
Comment (jfalcou) [inline test/unit/module/real/core/rotl.cpp:79]: remove outputs if not needed

**Review by jfalcou — CHANGES_REQUESTED** [review]

File: include/eve/function/rotl.hpp (hunk 1 above)
Comment (jtlap) [reply include/eve/function/rotl.hpp:30]: seemed clearer

File: include/eve/module/real/core/function/regular/generic/rotl.hpp (hunk 3 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/generic/rotl.hpp:28]: done

File: include/eve/module/real/core/function/regular/generic/rotl.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/generic/rotl.hpp:82]: don't know but not incorrect anyway

File: test/unit/module/real/core/rotl.cpp (hunk 6 above)
Comment (jtlap) [reply test/unit/module/real/core/rotl.cpp:79]: done

File: test/unit/module/real/core/rotl.cpp (hunk 5 above)
Comment (jtlap) [reply test/unit/module/real/core/rotl.cpp:64]: erroneous

File: include/eve/module/real/core/function/regular/generic/rotl.hpp (hunk 4 above)
Comment (jtlap) [reply include/eve/module/real/core/function/regular/generic/rotl.hpp:81]: removed

**Review by jfalcou — APPROVED** [review]

**Pull request merged**